README.rst
=======================
PyMessagePack pymsgpack
=======================

with the same performance of msgpack but also support python objects, sets, tuples.

msgpack protocol is modified so that there is no supports for other languages, except using packb with compatible_mode = true.

This is modified from msgpack-python (described below).

Here is the benchmark results on my x64 server, each for 200000 times:

::

   python performance.py 
   datatype    pymsgpack time(s)   cpickle time(s)  speed lifting ratio length(bytes)
   tuple0 ::    pymsgpack: 0.1      cpickle 0.173   ratio 1.73416934768 len:    3 : 4
   tuple1 ::    pymsgpack: 0.239    cpickle 0.487   ratio 2.03731597438 len:    255 : 507
   tuple2 ::    pymsgpack: 0.045    cpickle 0.121   ratio 2.64331974967 len:    804 : 822
   list0  ::    pymsgpack: 0.09     cpickle 0.208   ratio 2.29779431123 len:    1 : 6
   list1  ::    pymsgpack: 0.237    cpickle 0.518   ratio 2.18276936665 len:    253 : 508
   list11 ::    pymsgpack: 0.665    cpickle 2.719   ratio 4.08486282398 len:    708 : 1373
   list111::    pymsgpack: 0.206    cpickle 0.792   ratio 3.84661280845 len:    2253 : 2258
   list2  ::    pymsgpack: 0.083    cpickle 0.229   ratio 2.75877366213 len:    802 : 823
   dict0  ::    pymsgpack: 0.093    cpickle 0.205   ratio 2.18427884432 len:    1 : 6
   dict1  ::    pymsgpack: 0.041    cpickle 0.068   ratio 1.67492092969 len:    16 : 33
   dict2  ::    pymsgpack: 0.063    cpickle 0.108   ratio 1.71722933995 len:    1048 : 1070
   dict3  ::    pymsgpack: 0.651    cpickle 0.928   ratio 1.42595863402 len:    1669 : 1758
   set0   ::    pymsgpack: 0.112    cpickle 1.074   ratio 9.5092526728  len:    3 : 27
   set1   ::    pymsgpack: 0.084    cpickle 0.378   ratio 4.47175768422 len:    55 : 131
   set2   ::    pymsgpack: 0.052    cpickle 0.341   ratio 6.47297065153 len:    804 : 846
   newobj1::    pymsgpack: 0.307    cpickle 2.049   ratio 6.6596654085  len:    62 : 87
   oldobj1::    pymsgpack: 0.424    cpickle 1.39    ratio 3.27720864994 len:    62 : 87
   newobj2::    pymsgpack: 0.778    cpickle 3.213   ratio 4.12987933922 len:    7799 : 4618
   oldobj2::    pymsgpack: 0.858    cpickle 2.685   ratio 3.1270893371  len:    7796 : 4607


How to use:

1. git clone [this repo] and cd into it

2. make

3. sudo ./setup.py install

4. import pymsgpack in your script and use like msgpack

compatible_mode is available for communication with other msgpack interface.


=======================
MessagePack for Python
=======================

:author: INADA Naoki
:version: 0.4.6
:date: 2015-03-13

.. image:: https://secure.travis-ci.org/msgpack/msgpack-python.svg
   :target: https://travis-ci.org/#!/msgpack/msgpack-python
   

What's this
------------

`MessagePack <http://msgpack.org/>`_ is a fast, compact binary serialization format, suitable for
similar data to JSON. This package provides CPython bindings for reading and
writing MessagePack data.

Install
---------

::

   $ pip install msgpack-python

PyPy
^^^^^

msgpack-python provides pure python implementation.  PyPy can use this.

Windows
^^^^^^^

When you can't use binary distribution, you need to install Visual Studio
or Windows SDK on Windows.
Without extension, using pure python implementation on CPython runs slowly.

For Python 2.7, `Microsoft Visual C++ Compiler for Python 2.7 <https://www.microsoft.com/en-us/download/details.aspx?id=44266>`_
is recommended solution.

For Python 3.5, `Microsoft Visual Studio 2015 <https://www.visualstudio.com/en-us/products/vs-2015-product-editions.aspx>`_
Community Edition or Express Edition can be used to build extension module.


How to use
-----------

One-shot pack & unpack
^^^^^^^^^^^^^^^^^^^^^^

Use ``packb`` for packing and ``unpackb`` for unpacking.
msgpack provides ``dumps`` and ``loads`` as alias for compatibility with
``json`` and ``pickle``.

``pack`` and ``dump`` packs to file-like object.
``unpack`` and ``load`` unpacks from file-like object.

.. code-block:: pycon

   >>> import msgpack
   >>> msgpack.packb([1, 2, 3])
   '\x93\x01\x02\x03'
   >>> msgpack.unpackb(_)
   [1, 2, 3]

``unpack`` unpacks msgpack's array to Python's list, but can unpack to tuple:

.. code-block:: pycon

   >>> msgpack.unpackb(b'\x93\x01\x02\x03', use_list=False)
   (1, 2, 3)

You should always pass the ``use_list`` keyword argument. See performance issues relating to `use_list option`_ below.

Read the docstring for other options.


Streaming unpacking
^^^^^^^^^^^^^^^^^^^

``Unpacker`` is a "streaming unpacker". It unpacks multiple objects from one
stream (or from bytes provided through its ``feed`` method).

.. code-block:: python

   import msgpack
   from io import BytesIO

   buf = BytesIO()
   for i in range(100):
      buf.write(msgpack.packb(range(i)))

   buf.seek(0)

   unpacker = msgpack.Unpacker(buf)
   for unpacked in unpacker:
       print unpacked


Packing/unpacking of custom data type
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

It is also possible to pack/unpack custom data types. Here is an example for
``datetime.datetime``.

.. code-block:: python

    import datetime

    import msgpack

    useful_dict = {
        "id": 1,
        "created": datetime.datetime.now(),
    }

    def decode_datetime(obj):
        if b'__datetime__' in obj:
            obj = datetime.datetime.strptime(obj["as_str"], "%Y%m%dT%H:%M:%S.%f")
        return obj

    def encode_datetime(obj):
        if isinstance(obj, datetime.datetime):
            return {'__datetime__': True, 'as_str': obj.strftime("%Y%m%dT%H:%M:%S.%f")}
        return obj


    packed_dict = msgpack.packb(useful_dict, default=encode_datetime)
    this_dict_again = msgpack.unpackb(packed_dict, object_hook=decode_datetime)

``Unpacker``'s ``object_hook`` callback receives a dict; the
``object_pairs_hook`` callback may instead be used to receive a list of
key-value pairs.

Extended types
^^^^^^^^^^^^^^^

It is also possible to pack/unpack custom data types using the **ext** type.

.. code-block:: pycon

    >>> import msgpack
    >>> import array
    >>> def default(obj):
    ...     if isinstance(obj, array.array) and obj.typecode == 'd':
    ...         return msgpack.ExtType(42, obj.tostring())
    ...     raise TypeError("Unknown type: %r" % (obj,))
    ...
    >>> def ext_hook(code, data):
    ...     if code == 42:
    ...         a = array.array('d')
    ...         a.fromstring(data)
    ...         return a
    ...     return ExtType(code, data)
    ...
    >>> data = array.array('d', [1.2, 3.4])
    >>> packed = msgpack.packb(data, default=default)
    >>> unpacked = msgpack.unpackb(packed, ext_hook=ext_hook)
    >>> data == unpacked
    True


Advanced unpacking control
^^^^^^^^^^^^^^^^^^^^^^^^^^

As an alternative to iteration, ``Unpacker`` objects provide ``unpack``,
``skip``, ``read_array_header`` and ``read_map_header`` methods. The former two
read an entire message from the stream, respectively deserialising and returning
the result, or ignoring it. The latter two methods return the number of elements
in the upcoming container, so that each element in an array, or key-value pair
in a map, can be unpacked or skipped individually.

Each of these methods may optionally write the packed data it reads to a
callback function:

.. code-block:: python

    from io import BytesIO

    def distribute(unpacker, get_worker):
        nelems = unpacker.read_map_header()
        for i in range(nelems):
            # Select a worker for the given key
            key = unpacker.unpack()
            worker = get_worker(key)

            # Send the value as a packed message to worker
            bytestream = BytesIO()
            unpacker.skip(bytestream.write)
            worker.send(bytestream.getvalue())


Notes
-----

string and binary type
^^^^^^^^^^^^^^^^^^^^^^

In old days, msgpack doesn't distinguish string and binary types like Python 1.
The type for represent string and binary types is named **raw**.

msgpack can distinguish string and binary type for now.  But it is not like Python 2.
Python 2 added unicode string.  But msgpack renamed **raw** to **str** and added **bin** type.
It is because keep compatibility with data created by old libs. **raw** was used for text more than binary.

Currently, while msgpack-python supports new **bin** type, default setting doesn't use it and
decodes **raw** as `bytes` instead of `unicode` (`str` in Python 3).

You can change this by using `use_bin_type=True` option in Packer and `encoding="utf-8"` option in Unpacker.

.. code-block:: pycon

    >>> import msgpack
    >>> packed = msgpack.packb([b'spam', u'egg'], use_bin_type=True)
    >>> msgpack.unpackb(packed, encoding='utf-8')
    ['spam', u'egg']

ext type
^^^^^^^^

To use **ext** type, pass ``msgpack.ExtType`` object to packer.

.. code-block:: pycon

    >>> import msgpack
    >>> packed = msgpack.packb(msgpack.ExtType(42, b'xyzzy'))
    >>> msgpack.unpackb(packed)
    ExtType(code=42, data='xyzzy')

You can use it with ``default`` and ``ext_hook``. See below.

Note for msgpack-python 0.2.x users
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The msgpack-python 0.3 have some incompatible changes.

The default value of ``use_list`` keyword argument is ``True`` from 0.3.
You should pass the argument explicitly for backward compatibility.

`Unpacker.unpack()` and some unpack methods now raises `OutOfData`
instead of `StopIteration`.
`StopIteration` is used for iterator protocol only.

Note about performance
------------------------

GC
^^

CPython's GC starts when growing allocated object.
This means unpacking may cause useless GC.
You can use ``gc.disable()`` when unpacking large message.

use_list option
^^^^^^^^^^^^^^^^
List is the default sequence type of Python.
But tuple is lighter than list.
You can use ``use_list=False`` while unpacking when performance is important.

Python's dict can't use list as key and MessagePack allows array for key of mapping.
``use_list=False`` allows unpacking such message.
Another way to unpacking such object is using ``object_pairs_hook``.


Development
------------

Test
^^^^

MessagePack uses `pytest` for testing.
Run test with following command:

    $ py.test


..
    vim: filetype=rst


ChangeLog.rst
pymsgpack 0.1
=============
msgpack protocol is modified by our needs, and the never used tag 0xc1 is used for customized data 
such as set, tuple, and class instances (including new style and old style python class).

pymsgpack will temporarily not support other languages and will be used among same run time environments (with same import path and class definition)
the detailed reason for this limitation is that when we pack an instance, we extract the __dict__, __class__.__name__, __module__ of it,
and when we unpack the instance, we re-import the module and class, and build the instance using __new__, then set __dict__ to it.

the fallback supports is not available anymore.


----


msgpack 0.5.0
======

0.5 is important step toward 1.0.  There are some deprecations.
Please read changes carefully.

Changes
-------

* Drop Python 2.6 and 3.2 support

* Deprecate useless custom exceptions.  Use ValueError instead of PackValueError,
  Exception instead of PackException and UnpackException, etc...
  See msgpack/exceptions.py

* Add `strict_types` option to packer.  It can be used to serialize subclass of
  builtin types.  For example, when packing object which type is subclass of dict,
  `default()` is called.

* Pure Python implementation supports packing memoryview object.


0.4.8
=====
:release date: 2016-07-29

Bugs fixed
----------

* Calling ext_hook with wrong length. (Only on Windows, maybe. #203)


0.4.7
=====
:release date: 2016-01-25

Bugs fixed
----------

* Memory leak when unpack is failed

Changes
-------

* Reduce compiler warnings while building extension module
* unpack() now accepts ext_hook argument like Unpacker and unpackb()
* Update Cython version to 0.23.4
* default function is called when integer overflow


0.4.6
=====
:release date: 2015-03-13

Bugs fixed
----------

* fallback.Unpacker: Fix Data corruption when OutOfData.
  This bug only affects "Streaming unpacking."


0.4.5
=====
:release date: 2015-01-25

Incompatible Changes
--------------------

Changes
-------

Bugs fixed
----------

* Fix test failure on pytest 2.3.  (by @ktdreyer)
* Fix typos in ChangeLog.  (Thanks to @dmick)
* Improve README.rst  (by @msabramo)


0.4.4
=====
:release date: 2015-01-09

Incompatible Changes
--------------------

Changes
-------

Bugs fixed
----------

* Fix compile error.

0.4.3
=====
:release date: 2015-01-07

Incompatible Changes
--------------------

Changes
-------

Bugs fixed
----------

* Unpacker may unpack wrong uint32 value on 32bit or LLP64 environment. (#101)
* Build failed on Windows Python 2.7.

0.4.2
=====
:release date: 2014-03-26

Incompatible Changes
--------------------

Changes
-------

Bugs fixed
----------

* Unpacker doesn't increment refcount of ExtType hook.
* Packer raises no exception for inputs doesn't fit to msgpack format.

0.4.1
=====
:release date: 2014-02-17

Incompatible Changes
--------------------

Changes
-------

* fallback.Unpacker.feed() supports bytearray.

Bugs fixed
----------

* Unpacker doesn't increment refcount of hooks. Hooks may be GCed while unpacking.
* Unpacker may read unfilled internal buffer.

0.4.0
=====
:release date: 2013-10-21

Incompatible Changes
--------------------

* Raises TypeError instead of ValueError when packer receives unsupported type.

Changes
-------

* Support New msgpack spec.


0.3.0
=====

Incompatible Changes
--------------------

* Default value of ``use_list`` is ``True`` for now. (It was ``False`` for 0.2.x)
  You should pass it explicitly for compatibility to 0.2.x.
* `Unpacker.unpack()` and some unpack methods now raise `OutOfData` instead of
  `StopIteration`. `StopIteration` is used for iterator protocol only.

Changes
-------
* Pure Python fallback module is added. (thanks to bwesterb)
* Add ``.skip()`` method to ``Unpacker`` (thanks to jnothman)
* Add capturing feature. You can pass the writable object to
  ``Unpacker.unpack()`` as a second parameter.
* Add ``Packer.pack_array_header`` and ``Packer.pack_map_header``.
  These methods only pack header of each type.
* Add ``autoreset`` option to ``Packer`` (default: True).
  Packer doesn't return packed bytes and clear internal buffer.
* Add ``Packer.pack_map_pairs``. It packs sequence of pair to map type.



0.2.4
=======
:release date: 2012-12-22

Bugs fixed
----------

* Fix SEGV when object_hook or object_pairs_hook raise Exception. (#39)

0.2.3
=======
:release date: 2012-12-11

Changes
-------
* Warn when use_list is not specified. It's default value will be changed in 0.3.

Bugs fixed
-----------
* Can't pack subclass of dict.

0.2.2
=======
:release date: 2012-09-21

Changes
-------
* Add ``use_single_float`` option to ``Packer``. When it is true, packs float
  object in single precision format.

Bugs fixed
-----------
* ``unpack()`` didn't restores gc state when it called with gc disabled.
  ``unpack()`` doesn't control gc now instead of restoring gc state collectly.
  User can control gc state when gc cause performance issue.

* ``Unpacker``'s ``read_size`` option didn't used.

0.2.1
=======
:release date: 2012-08-20

Changes
-------
* Add ``max_buffer_size`` parameter to Unpacker. It limits internal buffer size
  and allows unpack data from untrusted source safely.

* Unpacker's buffer reallocation algorithm is less greedy now. It cause perforamce
  derease in rare case but memory efficient and don't allocate than ``max_buffer_size``.

Bugs fixed
----------
* Fix msgpack didn't work on SPARC Solaris. It was because choosing wrong byteorder
  on compilation time. Use ``sys.byteorder`` to get correct byte order.
  Very thanks to Chris Casey for giving test environment to me.


0.2.0
=======
:release date: 2012-06-27

Changes
-------
* Drop supporting Python 2.5 and unify tests for Py2 and Py3.
* Use new version of msgpack-c. It packs correctly on big endian platforms.
* Remove deprecated packs and unpacks API.

Bugs fixed
----------
* #8 Packing subclass of dict raises TypeError. (Thanks to Steeve Morin.)


0.1.13
=======

benchmarks/performance.py
#!/usr/bin/python2

from pymsgpack import packb, unpackb
from cPickle import dumps, loads



def allhash(content):
    all_hash = 0
    if isinstance(content, (str, int, float, tuple, bool)):
        all_hash += hash(content)
    elif isinstance(content, (dict, )):
        for k, v in content.iteritems():
            all_hash += allhash(k) + allhash(v)
    elif isinstance(content, (set, list)):
        for v in content:
            all_hash += allhash(v)
    elif hasattr(content, "__dict__"):
        all_hash += allhash(content.__dict__)
    return all_hash

class NewBase(object):
    def __init__(self):
        self.name = "NewBase"
        self.number = 12345678
        self.fnumber = 1.2345678

class NewMeddle1(NewBase):
    def __init__(self):
        super(NewMeddle1, self).__init__()
        self.name = "NewMiddle1"
        self.mt_1 = (1,2,3)
        self.t_000000000000000000000000000000000000000000000000000000 = ()
        self.l_000000000000000000000000000000000000000000000000000000 = []
        self.d_000000000000000000000000000000000000000000000000000000 = {}
        self.s_000000000000000000000000000000000000000000000000000000 = set()
        self.l_1 = [1, 1234567890, "", "-", "-"*8, "-"*256, "-"*512, 1.1]
        self.t_1 = tuple(self.l_1)
        self.s_1 = set(self.l_1)
        self.d_1 = {():[], (1,2,3):set([1,2,3]), 1:"1", "1":1, "-"*512:"="*512}

class NewMeddle2(NewBase):
    def __init__(self):
        super(NewMeddle2, self).__init__()
        self.name = "NewMiddle2"
        self.mt_2 = (4,5,6)
        self.i = 0

class NewChild(NewMeddle1, NewMeddle2):
    def __init__(self, i):
        super(NewChild, self).__init__()
        self.baseprop = NewBase()
        self.meddleprop1 = NewMeddle1()
        self.meddleprop2 = NewMeddle2()
        self.i = i
        self.name = "NewChild"

class OldBase:
    def __init__(self):
        self.name = "OldBase"
        self.number = 12345678
        self.fnumber = 1.2345678

class OldMeddle1(OldBase):
    def __init__(self):
        OldBase.__init__(self)
        self.name = "OldMiddle1"
        self.mt_1 = (1,2,3)
        self.t_000000000000000000000000000000000000000000000000000000 = ()
        self.l_000000000000000000000000000000000000000000000000000000 = []
        self.d_000000000000000000000000000000000000000000000000000000 = {}
        self.s_000000000000000000000000000000000000000000000000000000 = set()
        self.l_1 = [1, 1234567890, "", "-", "-"*8, "-"*256, "-"*512, 1.1]
        self.t_1 = tuple(self.l_1)
        self.s_1 = set(self.l_1)
        self.d_1 = {():[], (1,2,3):set([1,2,3]), 1:"1", "1":1, "-"*512:"="*512}

class OldMeddle2(OldBase):
    def __init__(self):
        OldBase.__init__(self)
        self.name = "OldMiddle2"
        self.mt_2 = (4,5,6)

class OldChild(OldMeddle1, OldMeddle2):
    def __init__(self, i):
        OldMeddle1.__init__(self)
        OldMeddle2.__init__(self)
        self.baseprop = OldBase()
        self.meddleprop1 = OldMeddle1()
        self.meddleprop2 = OldMeddle2()
        self.name = "OldChild"
        self.i = i

def pymsgpack_data(data):
    tp = packb(data)
    #print len(tp)
    tun = unpackb(tp)
    #print allhash(data), allhash(tun)
    #assert allhash(data) == allhash(tun)

def cpickle_data(data):
    tp = dumps(data, 2)
    #print len(tp)
    tun = loads(tp)
    #print allhash(data), allhash(tun)
    #assert allhash(data) == allhash(tun)

#import code
#code.interact(banner="", local=locals())

import time

def compare_all(times):
    somelist_0 = []
    somelist_1 = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50] * 5
    somelist_11 = [str(x) for x in somelist_1]
    somelist_111 = [x + 0.1 for x in somelist_1]
    somelist_2 = [1, 1234567890, "", "-", "-"*8, "-"*256, "-"*512, 1.1]
    sometuple_0 = tuple(somelist_0)
    sometuple_1 = tuple(somelist_1)
    sometuple_2 = tuple(somelist_2)
    someset_0 = set()
    someset_1 = set(somelist_1)
    someset_2 = set(somelist_2)
    somedict_0 = {}
    somedict_1 = {"1":1, "2":2, "3":3, "4":4, "5":5}
    somedict_2 = {2:(1,2,3,4,5,6,7), 1:(1,2,3), 1:"1", "1":1, "-"*512:"="*512}
    somedict_3 = {12032: 30, 12118: 0, 12033: 194, 10076: 0.0, 10071: 100, 12034: 395, 12121: 51, 12122: 0.04446481929146, 12123: 0.0, 10077: 0.0, 12124: 0.0, 12125: 0, 12128: 0.0, 12129: 0, 18000: 0, 18001: 0, 17001: 0, 17002: 0, 17003: 0, 17004: 0, 17005: 0, 17006: 0, 17007: 0, 17008: 0, 17009: 0, 17010: 0, 17011: 0, 17012: 0, 17013: 0, 17014: 0, 17015: 0.0, 17016: 0.0, 17017: 0.0, 17018: 0.0, 17019: 0.0, 17020: 0.0, 17021: 0.0, 16001: 75, 16002: 75, 16003: 0, 16004: 0, 16005: 0.5, 16006: 0.05, 16007: 0.0, 16008: 0.0, 16009: 0.0, 16010: 0, 16011: 0.0, 16012: 0.0, 16013: 0.0, 16014: 0.0, 16015: 0.0, 16016: 0.0, 16017: 0.0, 16018: 0, 16019: 90, 15001: 1.0, 15002: 1.0, 15003: 0.0, 15004: 0.0, 15005: 0.0, 15006: 0.0, 15007: 0.0, 15008: 0.0, 50010: 0, 14000: 0.0, 14001: 0.0, 14002: 0.0, 13001: 79, 13002: 158, 13003: 0.0, 13016: 0.0, 10066: 5, 12112: 0, 12010: 0, 12011: 0, 12012: 923, 12013: 1299, 12014: 0, 12015: 0, 12016: 416, 12017: 0.62783802048026, 12018: 408, 12019: 0.0, 12020: 174, 12021: 389, 12022: 0.0, 12023: 0.0, 11000: 7, 11001: 188, 11002: 218, 11003: 227, 11004: 13, 11005: 0, 11006: 0, 11007: 0, 11008: 0, 11009: 0, 11010: 0.0, 12035: 0.0, 12036: 0.0, 12037: 0.0, 12038: 0.0, 12039: 0.0, 12040: 341, 12041: 249, 12042: 0.0, 12113: 0.0, 10000: 1, 10001: 0, 10004: 0.0, 10005: 0.0, 10006: 0.0, 10007: 0.0, 10008: 96, 10009: 0, 52001: 0.0, 52002: 0.0, 52003: 0.0, 52004: 0.0, 52005: 0.0, 52006: 0.0, 52011: 0.0, 52012: 0.0, 52013: 0.0, 52014: 0.0, 52015: 0.0, 52016: 0.0, 51000: 4, 51001: 4, 51002: 4, 51003: 4, 51004: 4, 51005: 4, 51006: 4, 10050: 0, 10052: 0, 10053: 25797, 10054: 5917, 12103: 40, 12104: 0.33999999999999997, 12105: 451, 12106: 0.1, 10060: 0, 10061: 0.0, 10062: 25, 10063: 0.0, 10064: 1, 50001: 0, 12114: 42, 12115: 0.0, 12116: 766, 12117: 805, 10070: 0, 12119: 0.0, 12120: 0.7, 10073: 12.0, 10074: 5.0, 10075: 0.0, 50012: 0, 50013: 0, 12126: 0.0, 12127: 0, 10080: 337, 10081: 0, 12130: 0.0, 50019: 0, 50020: 2760, 50021: 0, 50022: 0, 10100: 0, 10101: 0, 50011: 2760, 10120: 3500, 10121: 3500, 10122: 0.0, 10123: 0.0, 12100: 2822, 10140: 0.0, 10141: 400.0, 10142: 0.0, 12101: 6601, 10144: 0, 10145: 0, 10146: 0, 10147: 0, 10148: 0, 10149: 0, 10150: 0.0, 10151: 300, 10152: 0, 10153: 300, 10154: 0, 10155: 0, 10143: 400.0, 12108: 0.0, 12109: 0.0, 12024: 0.01, 12025: 0.0, 12111: 0, 12026: 0.01482160592289, 50000: 1, 12027: 145, 10065: 0.0, 12028: 0.04, 50002: 0.0, 12029: 0.0, 10067: 0.0, 12030: 0.08, 12031: 0.0}
    litNew = NewBase()
    litOld = OldBase()
    bigNew = NewChild("bignew")
    bigOld = OldChild("bigold")

    compare_and_print(times * 5, sometuple_0,"tuple0")
    compare_and_print(times, sometuple_1,"tuple1")
    compare_and_print(times, sometuple_2,"tuple2")
    compare_and_print(times * 5, somelist_0, "list0 ")
    compare_and_print(times, somelist_1, "list1 ")
    compare_and_print(times, somelist_11, "list11 ")
    compare_and_print(times, somelist_111, "list111 ")
    compare_and_print(times, somelist_2, "list2 ")
    compare_and_print(times * 5, somedict_0, "dict0 ")
    compare_and_print(times, somedict_1, "dict1 ")
    compare_and_print(times, somedict_2, "dict2 ")
    compare_and_print(times, somedict_3, "dict3 ")
    compare_and_print(times * 5, someset_0,  "set0  ")
    compare_and_print(times, someset_1,  "set1  ")
    compare_and_print(times, someset_2,  "set2  ")
    compare_and_print(times * 5, litNew, "newobj1")
    compare_and_print(times * 5, litOld, "oldobj1")
    compare_and_print(times, bigNew, "newobj2")
    compare_and_print(times, bigOld, "oldobj2")

def compare_data(times, data):
    time1 = time.time()
    for j in xrange(times):
        pymsgpack_data(data)
    time2 = time.time()
    for j in xrange(times):
        cpickle_data(data)
    time3 = time.time()
    return time2 - time1, time3 - time2

def compare_and_print(times, data, name):
    t1, t2 = compare_data(times, data)
    len1 = len(packb(data))
    len2 = len(dumps(data, 2))
    print name, "::\tpymsgpack:", int(t1 * 1000)/1000.0, "\tcpickle", int(t2 * 1000)/1000.0, "\tratio", t2/t1, "len:\t", len1, ":", len2

if __name__ == "__main__":
    compare_all(80000)
